Add "Published" feature to dashboards

Tresdon opened the pull request:
This feature allows users to “publish” dashboards which essentially adds a notion of dashboards being public or private. The goal of this is to reduce the clutter that comes from a whole organization creating dashboards, many of which are not relevant to other users (or only a subset, whom they can share with selectively). The changes affect the following contexts:

On creating a dashboard:
	- Add a boolean published column to the dashboard model and allow the owners of a given dashboard to toggle this field in the UI.

On listing dashboards:
	- Don’t show all dashboards but instead only list dashboards with the following properties
		1. Dashboards the current user owns
		2. Dashboards that have been published (by any user)
		3. Dashboards which the user has favorited

Regarding point (3) above – The idea here was that no dashboard is strictly “private” in that it cannot be viewed if it is not published. This means that if a user creates an unpublished dashboard, they should still be able to send the link to others to share it with them. However, since the dashboard is unpublished the visitor wouldn’t see it in the list so they would need to keep the link around which isn’t a very friendly user experience. So to get around this we allow the user to favorite the dashboard and then it will show up in their list of dashboards.

issues which might be relevant to this: #1799 #2792 #4002

Tasks:
- [x] Add the 'published' column to the dashboard model and generate alembic migration file
- [x] Don't show unpublished dashboards on listing all dashboards.
- [x] Expose endpoint for toggling published status.
- [x] Protect published endpoint by allowing only owners to change the status.
- [x] Create React component on dashboards to represent and toggle status. 
- [x] Show error in the UI if somebody tries to toggle published on a dashboard which they don't own
- [x] Show Users favorited dashboards in the dashboard list even if they're not published
- [x] Write unit tests for dashboard visibility
- [x] Show interactions with the feature in this ticket

mistercrunch commented:
This is neat. A few thoughts:
* You probably didn't mean to publish `superset/assets/yarn.lock` as part of this PR
* My main concern is around launching this feature, all dashboards will suddenly become invisible, I'm not sure how to mitigate this, maybe a database script that looks for all active dashboards and set them as published? An active dashboard could be defined as one that a non-owner has accessed over the past 28 days or something?
* There needs to be an easy way for the owner to see the state `publish` / `not published` and to toggle it from the dashboard view itself, not from the CRUD view. We're trying to move away from our reliance on CRUD
* Similar feature could be even more relevant for Charts, where the clutter is much higher, maybe apply the same consistent pattern there?
* Curious to see what kind of SQL FAB generates when applying both `base_filters`, any way you can find out and share that? Let's make sure it's not doing anything crazy like running two SQL queries and showing the UNION. We should make sure this runs fast in envrionments where there are thousands of dashboards like at Airbnb.

Tresdon commented:
Thanks for such useful and quick feedback on this, A couple thoughts and things to do yet

-  Ah yeah the `yarn.lock` did creep in there. This is probably a silly question but what would the best way be to go about getting the proper version of this? Probably just using the one from the commit before or should redoing the `yarn install` set it straight?
- Releasing the feature: Perhaps it'd be fine to just have a script that sets all dashboards to published since this wouldn't affect the system at all and then people would just be able to modify it from then on? Even though 28 days-ish seems like a reasonable cutoff, this just seems minimally invasive.
- Toggling status in dashboard view and not just CRUD: Great idea, this commit adds some of that functionality. I've added an icon right next to the favorite icon to show whether or not it's published. I built it primarily off of the FavStar component as well. It's not currently working completely.
- About chart privacy: The charts do get more cluttered but I don't know if they're browsed as a final product in the same way as dashboards. Also wondering how this would interact with adding charts to dashboards, would people only be able to see published charts?
- SQL Generated: I'd be interested in knowing this too. I know that I can go to each of the `Filter`s and print out the query before and after any transformations but am not sure how to find out what things look like before applying either filter and afterwards. Any quick ideas on how to do this? I haven't looked into it much.

Things to do:
- The route I created for modifying the status `superset/dashboard/<id>/published/<action>` returns a 302. I haven't dug into it too much yet but perhaps you know what I need to do to address this.
- Show user their favorited dashboards in the list.
- Test non-owners trying to modify the published status
- Show generated SQL from using multiple base_filters

Thanks again for your feedback!

daniel5001 commented:
Hey guys,

Any update on this merge? Would love to be able to roll this feature out.

Thanks.

SpyderRivera commented:
I completely agree "Releasing the feature: Perhaps it'd be fine to just have a script that sets all dashboards to published since this wouldn't affect the system at all and then people would just be able to modify it from then on? "

Tresdon commented:
Hey @daniel5001 ,

This feature is still in the works and hasn't been forgotten, I'm currently working on ironing out a few bugs I've found as a result of adding an icon to toggle the status in the UI (and just trying to make it pleasant since it would show up on every dashboard) but other than that it's becoming pretty close to being finished!

Tresdon commented:
a quick visual tour of the things that this PR changes:

### Initial list of dashboards
![image](https://user-images.githubusercontent.com/9629830/39720733-8f858688-51fa-11e8-96c9-fe01ba76743e.png)

### After John publishes his secret dashboard

![image](https://user-images.githubusercontent.com/9629830/39720891-f0b34800-51fa-11e8-8c3a-941bd72813a5.png)

### Toggling the published status of a dashboard you own
![image](https://user-images.githubusercontent.com/9629830/39720947-226b507c-51fb-11e8-958e-ac89d364dd16.png)

### Trying to toggle the status of a dashboard you do not own
![image](https://user-images.githubusercontent.com/9629830/39720962-32ab1850-51fb-11e8-8951-dd3cbddf5a58.png)

Tresdon commented:
@mistercrunch This pull request should be finished now, let me know what you think. One thing I didn't mention was that this doesn't get rid of the `self.has_all_datasource_access()` check on dashboards and it still validates that user has access to all charts in the particular dashboard if they don't have access to all datasources. So right now for gamma users this is an additional constraint on top of them having access to all charts which we may want to change.

SpyderRivera commented:
LGTM

daniel5001 commented:
@mistercrunch  anything left to be done here? Looking forward to this feature. Thanks!

SpyderRivera commented:
![bitmoji](https://render.bitstrips.com/v2/cpanel/ee47f314-e678-4ac2-b3dc-5fa399e731c8-09178b39-9d8a-4dc6-94b9-2b76fa3848ae-v1.png?transparent=1&palette=1&width=246)

Tresdon commented:
@mistercrunch I've updated this feature to work with the dashboard improvements, anything else you'd like to see here?

SpyderRivera commented:
@Tresdon can you fix the dashboard tests issue?

michellethomas committed:
Allow users to publish dashboards

Tresdon commented:
Hey there @sylvia-tomiyama,

Thanks for taking a look at this and giving the user's experience some thought. I agree that some of the terminology surrounding the feature needs to be defined more clearly since there's some ambiguity around the term "published". You hit the nail on the head when you said that **My understanding is that unpublished dashboards are viewable, but less discoverable** – which is further bolstered by your distinction of _prod vs in-progress_. I think the comment about migration (which mentioned that dashboards would all of the sudden be gone) was referring to the visibility of them within these list views but they would not actually be gone or become truly unviewable. With that in mind I've considered some of the excellent points you've made and here's what I think:

### Welcome list & CRUD view list for dashboards
You mentioned as a part of your understanding that the visibility of the dashboard will change on the /welcome page but not the crud view. This wasn't quite my intention since when I was developing the feature these both used the same `/dashboardmodelview/list` endpoint and displayed the same dashboards. However I think it's a good idea to show admins all of the dashboards on the crud list view and to include a column of whether or not it's published for filtering. The alternative would make it so that even admins wouldn't be able to see these dashboards unless they performed a query which is less than ideal. 

Right now I say that only admins should be able to see all dashboards because I believe this view is used more frequently for finding dashboards than `/welcome` (considering I'm not aware of a way to return to `/welcome` except by setting the URL specifically). If the Superset icon in the top left returned the user to this view instead of the profile I would say it should affect just the welcome page and not the CRUD view but since most people use the CRUD view to find dashboards once they're in the application I think the feature should de-clutter that view as well. Perhaps the _profile_ button under the human icon in the top-right can become the `/profile` page and then an edit button on that page can take you to `/users/userinfo`? It seems like this may be more intuitive even without this feature we're discussing haha.

###   Displaying and Altering the Published Status
I think you make a great point about the eye icon being a bit misleading and a bit too subtle. I think a better solution might be something like a [_draft_] badge next to the title of the dashboard (possibly with a hover-over tooltip) which indicates a few things:

- The dashboard they're viewing will not show up in the list of dashboards, they should favorite it to see it in that list along with the others
- The elements used in the dashboard are tentative and subject to change
- Users should consider whether or not it is okay for them to share the dashboard before doing so

This indicator seems like it would make the intention of the feature much more understandable.
 
As for altering the status, I think you're totally right that it makes more sense to put this in the actions dropdown especially if we forgo the icon which was both the indicator and the toggle.

### The Way Forward
Here's a couple of action items which I think are appropriate after this discussion for me to work on:

- [ ] Replace the eye icon with a [draft] badge of sorts next to the title of unpublished dashboards.
- [ ] Add ability to toggle published status in the action dropdown menu 
- [ ] Add published field to CRUD listing of dashboards for sorting / filtering

Let me know what you think and thanks again for your consideration :)

Commit:
Merge branch 'master' of https://github.com/apache/incubator-superset into HEAD

Commit:
Rework publish dashboards feature

- The eye next to the title has been replaced with a [draft] badge
- Published status is now toggled in the Header Action Dropdown
- CRUD list shows published status

Commit:
Fix linter errors

Commit:
Update javascript tests

Commit:
Add tests and change DashboardFilter

Add some tests to make sure the published status is rendered and
Make it so that users cannot see dashboards that are published
if they don't have access to any of the slices within

Commit:
Fix some linter errors

Commit:
Merge master into publish_dashboards

Some tests are failing in travis CI and not locally so
I would like to see why

Commit:
Remove commas from core.py

Commit:
Merge test fixes from master

Commit:
Fix some failing tests

Commit:
Fix more linter errors I introduced

Commit:
More linter errors I introduced

elibrumbaugh commented:
@Tresdon Your walkthrough in Webex was super helpful. Would you be willing to share a screen recording with the group here?

Tresdon commented:
@elibrumbaugh, Certainly! Check out some of the recordings from above as well – it followed the idea of the toggle being front and center and also serving as the indicator which is something we discussed briefly on Friday.

## Toggling the Published Status of a Dashboard 
#### (The logged in user is "Admin User" in this example)

![image](https://user-images.githubusercontent.com/9629830/45964617-88e52480-bfe3-11e8-925c-f55ec2efb5a3.png)

Note a few things:
1. This user has the "all_datasource_access" permission so they can see all of the dashboards in the system.
2. One can only change the publish status by putting the dashboard into edit mode. 

## Not being able to toggle the status
#### (The logged in user is "Gamma User" in this example)

![image](https://user-images.githubusercontent.com/9629830/45964585-77038180-bfe3-11e8-94c6-e5616a7a06f6.png)

Note a few things:
1. The user cannot see the dashboard titled "No datasources for gamma" from the previous image. This is because this user doesn't have access to any of the datasources backing the slices used in the dashboard.
2. The user can see the dashboard they do because it is published AND they have access to at least one slice.
3. The option for publishing a dashboard does not appear for this user

Some of the design considerations we discussed are as follows:
- Publishing a dashboard is essentially the same as putting a dashboard into production right now so we need to make sure the visual weight of the action is reflected appropriately in the design of the feature
- This feature needs to be discoverable and have appropriate verbage that makes it clear how discoverable and visible the dashboard is. Perhaps public / private might be a more clear distinction for what this feature does.

Commit:
Merge branch 'master' of https://github.com/apache/incubator-superset into publish_dashboards

Commit:
Merge branch 'publish_dashboards' of https://github.com/Tresdon/incubator-superset into publish_dashboards

sylvia-tomiyama commented:
thanks for being so thoughtful in iterating with feedback! 

I think I understand a bit better what you're proposing "publish" does. To confirm, the current behavior/proposal is:

#### In welcome page and crud view
For admins
- show all dashboards regardless of status
- show a column that indicates the status of the dashboard (published, draft)

For everyone else
- show only public dashboards, and dashboards you own
- show a column that indicates the status of the dashboard (published, draft) //not sure if this was your intention, but I think this would make sense)

#### In profile page
- created content tab: show all dashboard you've created, regardless of status //it seems like it would be helpful to know the publication status here, too. like, "oh shit i forgot I was working on these, maybe I should publish them" / sanity check that you haven't published your work-in-progress stuff yet
- favorites tab: show favorited dashboards, regardless of publication status

#### On dashboard itself
For dashboard owner, for an unpublished dashboard
- show "draft" label next to title
- while in edit mode, you can select "Publish Dashboard" //nit: if we keep this, the D probably shouldn't be capitalized

For dashboard owner, for a published dashboard
- show nothing next to the title
- you can unpublish it

For everyone else
Unpublished dashboard
- show "draft" label next to title
- you can favorite it and it will appear in your favorites on your profile page
- can't edit status

Published dashboard
- show nothing
- you can favorite it and it will appear in your favorites on your profile page
- can't edit status

#### Feedback
### 
I think the "draft" label is a lot clearer than the eye icon. Although I originally suggested putting it in the dropdown, I think @elibrumbaugh has a good point that the corresponding publish action might be too buried in the drop-down and I like the idea of having it as its own separate button. I also think it could work as a toggle (where the indicator is also the button to flip), as long as it's not an eye icon (since we're intentionally making it visible even if it's not published). I think using words is also preferable to icons because then it's easy to map to the statuses that we'd show in the crud/welcome view, and I really like that you can understand the intention of the author of the dashboard as a consumer of a dashboard. I defer to @elibrumbaugh/design though.

With regards to terminology, I currently prefer published/draft because I think it feels closer to what the product behavior is. Public/private could be a little ambiguous (does public mean anyone can see it, even outside the company? Also if it were me, if it said private, I would assume that *no one* can view it, and that's not the case (and that's a bad mistake to make)).

sorab211 commented:
What are the plans for it to be merged to master?

Tresdon committed:
Merge branch 'master' into publish_dashboards

Commit:
update alembic migration

Commit:
Update design of publish dash feature

Commit:
Upgrade migration version

Commit:
Merge remote-tracking branch 'upstream/master' into publish_dashboards

Commit:
Secure publish endpoint

Commit:
Remove bad quotes

Commit:
Give publish span its own style

Commit:
fix publish rendering

Commit:
Add new test for publish feature

Commit:
Update migration

Commit:
update slug in test

pdoto commented:
Great work on this PR, looking forward to using it. hope it's gonna be merge to the master branch soon.

mistercrunch reviewed (changes requested):
Great work! I think this is really close. By default all users (including `all_datasource_access` ones) should see dashboards in the list that either:
* they own
* they favorited
* are published

Though maybe everyone should also be able to see all dashboards they have access to that are not published (other people's drafts) if they want to. Using the DashboardFilter approach is powerful (allows us to personalize the list), but it doesn't offer a way for users to opt out (as in show me all drafts).

There may be a way to introspect the `query` object to see if there's a filter on `published==False` and if that's the case, refrain from applying the `or_(published==True)`. Not a dealbreaker, but something to think about.

hughhhh commented on superset/views/core.py:
```diff
@@ -15,12 +15,13 @@
 from flask_appbuilder.actions import action
```
Could you move this its own .py file. `core.py` is getting really big and need to start breaking down the functionality.

Tresdon replied:
Responding to a similar comment below

hughhhh commented on superset/views/core.py:
```diff
@@ -2091,6 +2111,35 @@ def favstar(self, class_name, obj_id, action):
         session.commit()
         return json_success(json.dumps({'count': count}))
 
+    @api
```
Can we move this to a separate `.py` file. `core.py` is getting really big and should start trying to breakdown the endpoints into their separate file. 

@mistercrunch Can i get your opinion on this?

Tresdon replied:
I totally agree, `core.py` is kind of becoming a monolith which is less than ideal. In the same song I'm not sure if this is the feature to start breaking things out, it seems like it should be a whole refactoring effort so that concerns can be separated and it can be clearer what the PR is aiming at. For instance there's a `dashboard.py` file adjacent but that seems to only contain the logic for creating a new dashboard at this moment. I do think that everything with the name `Dashboard` should eventually move into that file (`DashboardModelView`, `DashboardAddView`, `DashboardFilter`, etc) but that seems like one rather self-contained task to me.

mistercrunch replied:
We recently started breaking apart the large modules into packages. In its current state, `core.py` stands for "views that haven't been sorted out yet".

The main challenge related to this is that the permission associated to the view need to be migrated properly. Anyhow, new dashboard-related views should be in `views/dashboard.py`, but we'll need an effort to move other views later regardless. 

File under "optional" for now

Commit:
Merge branch 'master' of https://github.com/apache/incubator-superset into publish_dashboards

Tresdon commented on superset/views/core.py:
```diff
@@ -94,36 +95,54 @@ def apply(self, query, func):  # noqa
+        3. Those which have been published (if they have access to at least one slice)
 
-    """List dashboards for which users have access to at least one slice or are owners"""
+    Show all dashboards for users which have the 'all_datasource_access' permission
```
Ahh interesting, I wasn't aware most folks were using superset this way – I'll have it just use the other rules instead. The concern here was that there should be a way for admins at least to see all the dashboards so that they may be able to purge ones that were left for dead unpublished without having to go into the DB and set the dashboard inactive there. I just think it's better to not assume that those with  the superset instance are necessarily DB-savvy

So if you think it's a fine idea, I will do a check for the current user having the admin role and then show them all the dashboards in that case only. The other option would be to encapsulate this in a "can_see_all_dashboards" permission in case some folks delete the admin role.

Commit:
Merge branch 'master' of https://github.com/apache/incubator-superset into publish_dashboards

Commit:
Update migration

Commit:
Address reviwer comments

Commit:
Fix linter errors

Commit:
Add licenses

Tresdon commented:
Thank you for the reviews @mistercrunch and @hughhhh! I've made some changes to address your comments – feel free to resolve the conversations if you think the concerns have been met :)

Commit:
Merge branch 'master' of https://github.com/apache/incubator-superset into publish_dashboards

Commit:
Remove fetchPublished(), use bootstrap data

Tresdon commented on superset/migrations/versions/a33a03f16c4a_add_extra_column_to_savedquery.py:
```diff
@@ -1,29 +1,30 @@
 # revision identifiers, used by Alembic.
 revision = 'a33a03f16c4a'
-down_revision = 'fb13d49b72f9'
+down_revision = '8b70aa3d0f87'
```
Sorry, was broke while I working on it so just wanted to see Travis happy but set it back

Tresdon commented on superset/assets/src/dashboard/actions/dashboardState.js:
```diff
@@ -110,6 +110,40 @@ export function saveFaveStar(id, isStarred) {
+  };
+}
+
+export function fetchPublished(id) {
```
I've made this change now, at first I was using the favorite star as a model but this is much better. it also resolves an issue where the badge would flash with the incorrect status before being set straight – a much better user experience!

Tresdon commented:
@mistercrunch another iteration of changes, lmk what you think!

SpyderRivera commented:
LGTM

nitin466 commented:
Great work. Is this PR merged yet ?

nitin466 commented:
Hi,
Can this "Publish feature" be used to handle following use case:
where I have four users with similar role(let's say Alpha) and i want to create dashboards which can be viewed by two alpha user(user1 and user2) but not to other users(user3 and user4) .  

Real use case: users with similar role but one or two attributes different. Suppose i have sales manager of two different geographical region and i want to create two separate dashboards corresponding to each region and each sales manager should be able to see dashboard of his region only not of others.

If Publish will not handle such cases, what are the other work around. Any suggestion ?
Thanks !

Tresdon commented:
Hello @nitin466,

If I understand you correctly the use case you're looking for is more granular (user-level instead of role-level) dashboard viewing permissions. This PR unfortunately does not address that use case, instead it allows for the publication of a dashboard across an organization.

I believe to accomplish the use case you highlighted above you will have to do a couple things:

- Create separate datasources for each geographic region
- Create geography-based roles and restrict the datasource permissions on those roles

Hopefully this can be of some help!

Commit:
Merge branch 'master' of https://github.com/apache/incubator-superset into publish_dashboards

Commit:
Update migration

Commit:
Update croniter to existing version

Commit:
Fix linter errors

Tresdon commented:
@mistercrunch anything else you would like to see done with this? The only thing I have been mulling over is giving the published and draft badges different background colors to help the user distinguish them from something other than the text itself. I like the idea of the draft being grey because of its associations with inactivity but perhaps the published badge could be "superset teal" (#00A699) or some more _active_ color. @elibrumbaugh , any thoughts on this?

mistercrunch commented:
This LGTM personally, I'd love to see another committer hit that merge button.

Tresdon commented:
Glad to hear it! Any eyes in particular you would like to check it out?

1AB9502 commented:
@mistercrunch ,

I would like to take over getting this story merged since @Tresdon got pulled into other things.
Would it be easier to open a new pull request, or figure out a way to resolve the merge conflicts in this pull request?

mistercrunch commented:
If you have write access to @Tresdon 's fork you could take over it. Otherwise you'll have to open a new PR off of your fork...

Tresdon commented:
Thanks for the assistance @1AB9502!

Commit:
Fix merge conflicts

Commit:
Merge upstream into branch

Commit:
Upgrade DB Revisions

Commit:
Fix flake8 linter error

Tresdon commented:
@mistercrunch @hughhhh 

I found some time to get this green again – I've also set up @1AB9502 with access to my fork so we can tag team the issue in the future.

One thing I'm wondering about is the process for releasing this feature – since it requires a db upgrade it could be that we want to set the `published` field as `true` for every dashboard at first so it seems that nothing has changed and everything is still visible. This seems like it might be the most accessible UX for folks at first but I'm open to other ideas. If you're in agreement I'm wondering what the best way to accomplish this would be. I'm thinking of just a simple `UPDATE table` statement but not sure this is the most elegant or appropriate way.

mistercrunch reviewed (changes requested):
@Tresdon just one thing, let's set the database migration to set all existing dashboards to `published = True`

We could also set a feature flag in front of the feature if people feel strongly about NOT enabling this feature in their environment, but I doubt this would be the case.

Commit:
Set all dashboards to published on migration

Commit:
Merge branch 'master' of https://github.com/apache/incubator-superset into publish_dashboards

Commit:
Migration proper line spacing

mistercrunch commented:
Hey sorry about the new conflicts, but they should be easy to solve, we just started using `black`,  so you should be able to just run `black superset/` and `black tests/`, and then rebase and things should go smooth. Let us know if that's not the case...

https://github.com/python/black

Commit:
Fix migration to work with postgres

Commit:
Use black code formatting

Tresdon commented:
Zero worries, I think the formatting as well as the migration initialization are both figured out – but waiting to hear if travis has anything to say about linting. I quite like the feature flag system for the project and am willing to implement it – but I also don't foresee many folks finding that it's a feature they can't live with.

Commit:
UPDATE statement works with postgresql and sqlite hopefully

Tresdon commented:
@mistercrunch Any idea how to re-run travis without changing files? The error it got seems to be a `Connection reset by peer`. The test failed on making a request to `https.github.com` and I can't seem to reproduce it locally

Commit:
Merge branch 'master' of https://github.com/apache/incubator-superset into publish_dashboards

Commit:
Update wording to kick off travis

Tresdon commented:
@mistercrunch any thoughts on the changes or does it all look good to you?
